### DISCUSSION TITLE
터놓고 인프라 버전업 내용

### DISCUSSION THREAD

**Discussion by her0807** [woowacourse-teams/2022-ternoko#438] [Announcements]
https://github.com/woowacourse/infra-architecture-4/discussions/2#discussioncomment-3792280

@soominsohn  뚜민아 여기에도 내용 넣어줘

**Comment by soominsohn** [comment]
## 터놓고
**BE** 
열음 애쉬 바니 수달 앤지


**기존 터놓고 인프라 아키텍처**

<img width="532" alt="image" src="https://user-images.githubusercontent.com/83059234/193723820-e598fd1e-65fa-44fa-b54b-8b3bde0e512d.png">

1. NGINX 꺼지면?
    1. nginx가 죽은걸 어떻게 파악할 수 있을까?
    2. 헬스체크 스크립트를 작성한다.
    3. NGINX 두대를 놓는다.
        1. 그러면 요청 분산 어떻게 할건지?
    4. 포트포워딩으로 관리한다.
    5. 밖에서 엔진엔스 헬스체크 해준다.
        1. 포트가 떠있는지 체크해줌.
        2. 죽으면 다른 포트로 띄우겠다.
    6. 다른 리전에다가 새로 띄운다.
    7. nginx dashboard가 있습니다.

1. WAS가 꺼지면?
    1. WAS를 여러대 두어서 로드밸런싱 도입.
    2. nginx가 처리해준다~
    3. WAS끼리 동기화는 어떻게 해야할까?
        1. DB에 낙관락을 건다.

1. DB가 꺼지면?
    1. 백업 DB를 둔다.
    2. Replication DB를 둔다.
        - 단방향 Replication: 읽기전용
        - 양방향 Replication: 데이터 동기화 가능. 읽기/쓰기 둘 다 가능
        - 읽기전용 replication db를 두고 master db가 죽으면?
            - 백업전용 replication db를 두고 백업을 한다.
            - replica 서버 master로 승격시키고, 새로 slave를 만든다.
    3. DB가 죽은걸 어떻게 알 수 있을까?
        1. ping을 보낸다.
    4. MySQL 헬스체크
    5. 샤딩 파티셔닝 등등…

1. EC2가 꺼지면?
    1. 새로 띄운다.
    2. 도커 이미지를 사용한다.
        1. 이미지 어떻게 뜨냐?
        2. 이미지 떠서 도커허브(깃헙 같은거)에 올린다.
        3. 허브에서 이미지 땡겨와서 올린다.

1. AWS에 모니터링 하는 것 없나?
    1. 각자 헬스체크하는 기능이 없나?
    2. 모니터링하는 기술이 있을 것 같다.

1. FE도 여러대 두자.
    1. FE 앞에 nginx 두고 FE WS 여러대 두자.
    2. 그렇다면 FE는 어떤 무중단 배포를 사용해야 할까?
    3. FE - 카나리 배포, BE - 블루그린 배포를 도입했을 때
        1. FE와 BE가 동기화 되지 않은 상황이 발생할 수 있다.
        2. 브라우저 캐시문제도 있다. 
            1. 배포를 할 때 브라우저 캐시를 무효화하는 로직을 넣는다.
    4. 사이에 들어오는 요청을 에러처리 하자.

1. 솔루션
    1. 장애를 파악하는 툴이 있다. (대시보드 같은 것) (nginx로 가능할 듯)
        1. 예) WAS 164 꺼졌어요! 재부팅 해본다.
        2. 그런데 안되면?
        3. 개발자 불러서 장애 복구한다.
    2. 근데 nginx가 꺼지면?
        1. 자동화가 가능할까?
        2. 모니터링 툴을 도입한다.
            1. ec2 헬스체크
            2. log 분석
    3. 모든 것을 자동화 할 수 없다면 사람의 개입을 최소화하는 방향으로 개선한다.
    

**개선해 본 점**

<img width="567" alt="image" src="https://user-images.githubusercontent.com/83059234/193723849-acdb1800-05e2-472e-a626-ecd5b63fb874.png">
